src/Transport/ExtensionSnmp.php

```diff
@@ -73,56 +81,43 @@ public function __construct(
         string $privPassphrase = 'None'
     ) {
         $this->community = $community;
-        $this->host      = $host;
-        $this->retry     = $retry;
-        $this->timeout   = $timeout;
-        $this->version   = $version;
-
-        $this->secName        = $community;
-        $this->secLevel       = $secLevel;
-        $this->authProtocol   = $authProtocol;
+        $this->host = $host;
+        $this->retry = $retry;
+        $this->timeout = $timeout;
+        $this->version = $version;
+
+        $this->secName = $community;
+        $this->secLevel = $secLevel;
+        $this->authProtocol = $authProtocol;
         $this->authPassphrase = $authPassphrase;
-        $this->privProtocol   = $privProtocol;
+        $this->privProtocol = $privProtocol;
         $this->privPassphrase = $privPassphrase;
         snmp_set_oid_output_format(SNMP_OID_OUTPUT_NUMERIC);
     }
 
     /**
-     * @return mixed[]
+     * @return iterable<string, mixed>
      */
-    public function walkFirstDegree(string $walkedOid) : iterable
+    public function walk(string $oid) : iterable
     {
-        $result = $this->realWalk($walkedOid);
-
-        $oidPrefix = null;
-        foreach ($result as $oid => $value) {
-            $length = strrpos($oid, '.');
-            assert(is_int($length));
-
-            if ($oidPrefix !== null && $oidPrefix !== substr($oid, 0, $length)) {
-                throw new SnmpFailed('Requested OID tree is not a first degree indexed SNMP value');
-            }
-
-            $oidPrefix = substr($oid, 0, $length);
-
-            yield substr($oid, $length + 1) => $this->parseSnmpValue($value);
+        $oidLength = strlen($oid) + 1;
+        foreach ($this->realWalk($oid) as $oidKey => $value) {
+            yield substr($oidKey, $oidLength) => $this->parseSnmpValue($value);
         }
     }
 
     /**
-     * @return mixed[]
+     * @return iterable<string, mixed>
      */
-    public function walk(string $oid) : iterable
+    public function walkWithCompleteOids(string $oid) : iterable
     {
-        $rawResult = $this->realWalk($oid);
-
-        foreach ($rawResult as $oidKey => $value) {
+        foreach ($this->realWalk($oid) as $oidKey => $value) {
             yield $oidKey => $this->parseSnmpValue($value);
         }
     }
 
     /**
-     * @return mixed[]
+     * @return array<string, mixed>
```

Comment:
```suggestion
     * @return iterable<string, mixed>
```

Reply:
and array is iterable 🤔

Reply:
and `RecursiveCallbackFilterIterator` is also `object`, and yet you wouldn't use `object` as a return type for a method returning an instance of `RecursiveCallbackFilterIterator` :)

Reply:
yes but iterable is IMO safer in most cases as array often gets abused in PHP. And the proof here is that result is only iterated so you don't need array access anyway. But your call, the method is private.

Reply:
When accepting values, you must be as broad as possible. On the other hand, when returning values, you must be as strict as possible. Narrowing/broadening the types doesn't help, (afaik) ever.